### PR TITLE
Fix on cart page+  new design

### LINKED ISSUE #13229: Promo code during checkout and order button
**Describe the bug**
1) Try to add a promo code during checkout, try to open and close it
The effect is jerky

2) The "order" button is stuck to the frame

capture d'écran_1323

3) When you have promo code highlighted and you arrive on the checkout you have 2 links: "Have a promo code ?" and "Close" to do the same thing
"Have a promo code?" shouldn't be displayed in this case

capture d'écran_1331

4) When you click on close the promo code higlighted is closed, it should be only the promo field

**Additionnal information**
PrestaShop version: develop
PHP version: N/A

### PR DESCRIPTION
**Pull request by YeLnatSs**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       |  1.7.6.x
| Description?  | Fix on cart page and new design
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | Does it break backward compatibility? no
| Deprecations? | Does it deprecate an existing feature? no
| Fixed ticket? | Fixes #13229 
| How to test?  | Add a product to your cart and click on "proceed to checkout" to go on cart page.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/13553)

### CODE REVIEW

File: themes/classic/_dev/js/cart.js [hunk 1]
```diff
@@ -168,6 +168,36 @@ $(document).ready(() => {
...
+
+  function labelCodeDisplayed() {
+    $('.promo-code-button').not('.cancel-promo').hide();
+    
+    $('.cancel-promo').on('click', () => {
+      $('.cancel-promo').hide();
+      $('.promo-code-button').not('.cancel-promo').show();
+    });
+    $('.promo-code-button').not('.cancel-promo').on('click', () => {
+        $('.promo-code-button').not('.cancel-promo').hide();
```
Comment (Quetzacoalt91) [inline themes/classic/_dev/js/cart.js]: $(this) should be working, isn't it?
```suggestion
        $(this).hide();
```

File: themes/classic/_dev/js/cart.js [hunk 2]
```diff
@@ -168,6 +168,36 @@ $(document).ready(() => {
...
 
+  //Hide 'have a code promo' text if displayed when 'close' is
+  let isPresent = $('.promo-code-button').length,
+      nameLength = 0;
+
+  function labelCodeDisplayed() {
+    $('.promo-code-button').not('.cancel-promo').hide();
+    
+    $('.cancel-promo').on('click', () => {
+      $('.cancel-promo').hide();
```
Comment (Quetzacoalt91) [inline themes/classic/_dev/js/cart.js]:
```suggestion
      $(this).hide();
```

should be working, and prevent a lookup in the DOM

File: themes/classic/_dev/js/cart.js [hunk 3]
```diff
@@ -168,6 +168,36 @@ $(document).ready(() => {
...
+    });
+    $('.promo-code-button').not('.cancel-promo').on('click', () => {
+        $('.promo-code-button').not('.cancel-promo').hide();
+        $('.cancel-promo').show();
+    });
+  }
+
+  setInterval( function () {
+    if ($('.promo-name').length != nameLength){
+      nameLength = $('.promo-name').length;
```
Comment (Quetzacoalt91) [inline themes/classic/_dev/js/cart.js]: `.length` is not chached by the javascript engine. It would be great to store this value in a variable

File: themes/classic/_dev/js/cart.js (hunk 1 above)
Comment (PierreRambaud) [reply themes/classic/_dev/js/cart.js]: No because of es6 notation `() => {} `

it should be 
```js
$('.promo-code-button').not('.cancel-promo').on('click', function hideCancePromo() {
        $(this).hide();
}):
```

File: themes/classic/_dev/js/cart.js [hunk 4]
```diff
@@ -168,6 +168,39 @@ $(document).ready(() => {
...
+    });
+
+    $('.promo-code-button').not('.cancel-promo').on('click', (e) => {
+        $(e.currentTarget).hide();
+        $('.cancel-promo').show();
+    });
+  }
+
+  setInterval( function () {
+    let promoNameLength = $('.promo-name').length;
```
Comment (PierreRambaud) [inline themes/classic/_dev/js/cart.js]:
```suggestion
    const promoNameLength = $('.promo-name').length;
```

File: themes/classic/_dev/js/cart.js [hunk 5]
```diff
@@ -168,6 +168,39 @@ $(document).ready(() => {
...
+
+    if (promoNameLength != nameLength) {
+      nameLength = promoNameLength;
+      labelCodeDisplayed();
+    }
+  }, 100);
+
+  $('.promo-code-button').each( () => {
```
Comment (PierreRambaud) [inline themes/classic/_dev/js/cart.js]:
```suggestion
  $('.promo-code-button').each(() => {
```

File: themes/classic/_dev/js/cart.js [hunk 6]
```diff
@@ -168,6 +168,39 @@ $(document).ready(() => {
...
+
+  function labelCodeDisplayed() {
+    $('.promo-code-button').not('.cancel-promo').hide();
+    
+    $('.cancel-promo').on('click',  (e) => {
+      $(e.currentTarget).hide();
+      $('.promo-code-button').not('.cancel-promo').show();
+    });
+
+    $('.promo-code-button').not('.cancel-promo').on('click', (e) => {
```
Comment (PierreRambaud) [inline themes/classic/_dev/js/cart.js]:
```suggestion
    $('.promo-code-button:not(.cancel-promo)').on('click', (e) => {
```

File: themes/classic/_dev/js/cart.js [hunk 7]
```diff
@@ -168,6 +168,39 @@ $(document).ready(() => {
     });
   };
 
+  //Hide 'have a code promo' text if displayed when 'close' is
+  let isPresent = $('.promo-code-button').length,
+      nameLength = 0;
+
+  function labelCodeDisplayed() {
+    $('.promo-code-button').not('.cancel-promo').hide();
```
Comment (PierreRambaud) [inline themes/classic/_dev/js/cart.js]:
```suggestion
    $('.promo-code-button:not(.cancel-promo)').hide();
```

File: themes/classic/_dev/js/cart.js [hunk 8]
```diff
@@ -168,6 +168,39 @@ $(document).ready(() => {
...
+    
+    $('.cancel-promo').on('click',  (e) => {
```
Comment (rdy4ever) [inline themes/classic/_dev/js/cart.js]: Shouldn't all classes used by js have a `js-` prefix? I see a lot of javascript executed on common class names. Wouldn't this break stuff for child themes? Keeping non `js-` prefixed classes unchanged is not mandatory and some developers may have removed or changed class names in their templates.

Comment (PierreRambaud) [reply themes/classic/_dev/js/cart.js]: I'm really not in favor to add a `js-` for class with js only, what happened if we improve theme and use a css on this class? More if a child theme use this `js-` for a css usage.
Keep thinking that a class is like an id, just a selector, not only for css.
It's really overkill and add characters for nothing :/

File: themes/classic/_dev/js/cart.js (hunk 1 above)
Comment (Quetzacoalt91) [reply themes/classic/_dev/js/cart.js]: Interesting. Learnt something one more time.

File: themes/classic/_dev/js/cart.js [hunk 9]
```diff
@@ -168,6 +168,39 @@ $(document).ready(() => {
...
+  setInterval( function () {
+    const promoNameLength = $('.promo-name').length;
+
+    if (promoNameLength != nameLength) {
+      nameLength = promoNameLength;
+      labelCodeDisplayed();
+    }
+  }, 100);
+
+  $('.promo-code-button').each(() => {
```
Comment (Quetzacoalt91) [inline themes/classic/_dev/js/cart.js]: Why is this line needed?

File: themes/classic/_dev/js/cart.js (hunk 1 above)
Comment (PierreRambaud) [reply themes/classic/_dev/js/cart.js]: Yes, with `() => {}` `this  `context came from parent / class instead of `function` which have is own `this` context 😄

File: themes/classic/_dev/js/cart.js (hunk 9 above)
Comment (YeLnatSs) [reply themes/classic/_dev/js/cart.js]: at least that one is useless. I remove it

File: themes/classic/templates/checkout/_partials/cart-voucher.tpl [hunk 10]
```diff
@@ -60,22 +66,31 @@
...
+          aria-controls="promo-code">
+            {l s='Close' d='Shop.Theme.Checkout'}
+          </a>
+        </p>
+
+        {if $cart.discounts|count > 0}
+          <p class="block-promo promo-highlighted">
+            {l s='Take advantage of our exclusive offers:' d='Shop.Theme.Actions'}
+          </p>
+        {/if}
```
Comment (Quetzacoalt91) [inline themes/classic/templates/checkout/_partials/cart-voucher.tpl]: So from now you always display the `promo-discounts`, even empty?

File: themes/classic/_dev/js/cart.js [hunk 11]
```diff
@@ -168,6 +168,49 @@ $(document).ready(() => {
     });
   };
 
+  //Hide 'have a code promo' text if displayed when 'close' is
+  let isPresent = $('.promo-code-button').length,
```
Comment (jolelievre) [inline themes/classic/_dev/js/cart.js]: If `isPresent` is not a boolean but a test on the number of buttons then you should rename it, for example `promoButtonsNb`

File: themes/classic/_dev/js/cart.js [hunk 12]
```diff
@@ -168,6 +168,49 @@ $(document).ready(() => {
...
+  }
+
+  function isBlockPromoPresent() {
+    if ($('.block-promo-container').length != 0 && isPresent === 2) {
+      $('.cancel-promo').show();
+      $('.promo-code-button').not('.cancel-promo').hide();
+    }
+  }
+
+  setInterval( function () {
```
Comment (jolelievre) [inline themes/classic/_dev/js/cart.js]: Be careful `setInterval` will happen every 100ms not only once, so `labelCodeDisplayed` will be called several times (which is not useful, once is enough) and it is really bad for performance
If you only need this code to execute once, you better use a setTimeout or at least clear your interval when it's not needed any more

I'm not sure what is the purpose of this code? hide the buttons when the user has typed the right amount of characters?

File: themes/classic/_dev/js/cart.js [hunk 13]
```diff
@@ -168,6 +168,49 @@ $(document).ready(() => {
     });
   };
 
+  //Hide 'have a code promo' text if displayed when 'close' is
+  let isPresent = $('.promo-code-button').length,
+      nameLength = 0;
+  
+  $('.cancel-promo').hide();
+
+  function labelCodeDisplayed() {
```
Comment (jolelievre) [inline themes/classic/_dev/js/cart.js]: The function name suggest a test returning a boolean, you should rename it
`initPromoCodeButtons` for example
You only hide buttons? not the input?

File: themes/classic/_dev/js/cart.js [hunk 14]
```diff
@@ -168,6 +168,49 @@ $(document).ready(() => {
...
+      $('.promo-code-button:not(.cancel-promo)').show();
+    });
+
+    $('.promo-code-button:not(.cancel-promo)').on('click', (e) => {
+        $(e.currentTarget).hide();
+        $('.cancel-promo').show();
+    });
+  }
+
+  function isBlockPromoPresent() {
```
Comment (jolelievre) [inline themes/classic/_dev/js/cart.js]: The function name suggest a test returning a boolean, you should rename it
`initPromoMode` or `forceCancelButtonOnPromo` maybe? I'm not sure of this use case and what you are aiming
Also one selector should ne enough `$('.promo-code-button:not(.cancel-promo)').hide();`

File: themes/classic/_dev/js/cart.js [hunk 15]
```diff
@@ -168,6 +168,49 @@ $(document).ready(() => {
     });
   };
 
+  //Hide 'have a code promo' text if displayed when 'close' is
+  let isPresent = $('.promo-code-button').length,
+      nameLength = 0;
+  
+  $('.cancel-promo').hide();
```
Comment (jolelievre) [inline themes/classic/_dev/js/cart.js]: If this is the default state you should use css to hide it by default

**Review by jolelievre — CHANGES_REQUESTED** [review]

File: themes/classic/_dev/js/cart.js (hunk 11 above)
Comment (YeLnatSs) [reply themes/classic/_dev/js/cart.js]: No problem, I changed it to your proposition

File: themes/classic/_dev/js/cart.js (hunk 12 above)
Comment (YeLnatSs) [reply themes/classic/_dev/js/cart.js]: Okay, thank you.
It was when you have a promo code and add it, "have a promo code" and "close" were displayed at the same time. So I add this to manage it.
The promo code is displayed from a condition in tpl file.

File: themes/classic/_dev/js/cart.js (hunk 13 above)
Comment (YeLnatSs) [reply themes/classic/_dev/js/cart.js]: Ok, thanks. I just change it.
Yes only buttons.

File: themes/classic/_dev/js/cart.js (hunk 14 above)
Comment (YeLnatSs) [reply themes/classic/_dev/js/cart.js]: I changed it to 'forceCancelButtonOnPromo' 😉

File: themes/classic/_dev/js/cart.js (hunk 15 above)
Comment (YeLnatSs) [reply themes/classic/_dev/js/cart.js]: Yes, you're right

File: themes/classic/_dev/js/cart.js (hunk 11 above)
Comment (PierreRambaud) [reply themes/classic/_dev/js/cart.js]: As I know, we can have only one promo code button :thinking: 
@jolelievre The length is the only method to tests if a selector exists in jQuery
The `isPresent` was legit :)

Comment (jolelievre) [reply themes/classic/_dev/js/cart.js]: No if you look at the code after there is a test `isPresent == 2`
There can be two buttons one for showing the other for canceling

File: themes/classic/_dev/js/cart.js (hunk 12 above)
Comment (jolelievre) [reply themes/classic/_dev/js/cart.js]: Yes but you could it with the onReady event, no need for a setInterval

File: themes/classic/_dev/js/cart.js [hunk 16]
```diff
@@ -169,27 +169,25 @@ $(document).ready(() => {
...
+  function initPromoCodeButtons() {
     $('.cancel-promo').on('click', (e) => {
       $(e.currentTarget).hide();
       $('.promo-code-button:not(.cancel-promo)').show();
     });
 
     $('.promo-code-button:not(.cancel-promo)').on('click', (e) => {
         $(e.currentTarget).hide();
-        $('.cancel-promo').show();
+        $('.cancel-promo').css('display','block');
```
Comment (jolelievre) [inline themes/classic/_dev/js/cart.js:183]: Isn't it what `show()` does?

Comment (YeLnatSs) [reply themes/classic/_dev/js/cart.js:183]: Yes but the target is an inline element so I had to force with 'display: block' to make it look like it was before

File: themes/classic/_dev/js/cart.js (hunk 11 above)
Comment (PierreRambaud) [reply themes/classic/_dev/js/cart.js]: Lol, so yes, it's not relevant ^^

File: themes/classic/_dev/js/cart.js (hunk 12 above)
Comment (jolelievre) [reply themes/classic/_dev/js/cart.js]: I still don't understand the purpose of `setInterval` ?

Comment (YeLnatSs) [reply themes/classic/_dev/js/cart.js]: I put a setInterval because when you have a promo code and add it, both texts "have a promo code" and "closed" were displayed. Normally, only "have a promo code." is supposed to be displayed once you add your promo code.
'setInterval' was there to call again my function to fix that bug once you add a promo code.
The promo code is displayed from a condition in tpl file '{if $cart.vouchers.added}'.

Comment (jolelievre) [reply themes/classic/_dev/js/cart.js]: Well then you shouldn't use `setInterval` this forces an useless overload every 100ms you perform this check just in case the promo code has been added

If this code is supposed to be executed when a promo code is added, I suppose it is added via javascript then you have to display your buttons correctly at this moment
If it is displayed thanks to smarty in a tpl file then it should already be on your page no? Unless it is updated via ajax?

File: themes/classic/_dev/js/cart.js [hunk 17]
```diff
@@ -168,6 +168,47 @@ $(document).ready(() => {
...
+  }
+
+  function forceCancelButtonOnPromo() {
+    if ($('.block-promo-container').length != 0 && promoButtonsNb === 2) {
+      $('.cancel-promo').css('display','block');
+      $('.promo-code-button:not(.cancel-promo)').hide();
+    }
+  }
+
+  setTimeout( function () {
```
Comment (jolelievre) [inline themes/classic/_dev/js/cart.js:194]: I still don't understand why you need this `setTimeout` you are already in a `$(document).ready` code block so the DOM should be loaded already.

**Comment by PierreRambaud** [comment]
Closed in favor of #14180
Thanks @YeLnatSs
